src/librustc_mir/transform/check_consts/qualifs.rs

```diff
@@ -0,0 +1,290 @@
+use rustc::mir::*;
+use rustc::mir::interpret::ConstValue;
+use rustc::ty::{self, Ty};
+use rustc_data_structures::bit_set::BitSet;
+use syntax_pos::DUMMY_SP;
+
+use super::Item as ConstCx;
+use super::validation::Mode;
+
+#[derive(Clone, Copy)]
+pub struct QualifSet(u8);
+
+impl QualifSet {
+    fn contains<Q: ?Sized + Qualif>(self) -> bool {
+        self.0 & (1 << Q::IDX) != 0
+    }
+}
+
+/// A "qualif"(-ication) is a way to look for something "bad" in the MIR that would disqualify some
+/// code for promotion or prevent it from evaluating at compile time. So `return true` means
+/// "I found something bad, no reason to go on searching". `false` is only returned if we
+/// definitely cannot find anything bad anywhere.
+///
+/// The default implementations proceed structurally.
+pub trait Qualif {
+    const IDX: usize;
+
+    /// Whether this `Qualif` is cleared when a local is moved from.
+    const IS_CLEARED_ON_MOVE: bool = false;
+
+    /// Return the qualification that is (conservatively) correct for any value
+    /// of the type, or `None` if the qualification is not value/type-based.
+    fn in_any_value_of_ty(_cx: &ConstCx<'_, 'tcx>, _ty: Ty<'tcx>) -> Option<bool> {
+        None
+    }
+
+    /// Return a mask for the qualification, given a type. This is `false` iff
+    /// no value of that type can have the qualification.
+    fn mask_for_ty(cx: &ConstCx<'_, 'tcx>, ty: Ty<'tcx>) -> bool {
+        Self::in_any_value_of_ty(cx, ty).unwrap_or(true)
+    }
+
+    fn in_static(_cx: &ConstCx<'_, 'tcx>, _static: &Static<'tcx>) -> bool {
+        // FIXME(eddyb) should we do anything here for value properties?
+        false
```

Comment:
Okay, so use `in_any_value_of_ty` here.

Reply:
See ac7a343.

Reply:
Changing this [broke some tests](https://github.com/rust-lang/rust/pull/64470#issuecomment-535231283), notably [this one](https://github.com/rust-lang/rust/blob/6c2c29c43206d6e2f1091fa278d2792ea10e3659/src/test/ui/consts/std/cell.rs#L10). It seems like we should explicitly allow references to `HasInteriorMut` locals in `static`s rather than claiming that `static`s never have any qualification.

In any case, I think this qualifies as a non-trivial change to the existing logic and should go in another PR.

Reply:
I think all you need to do is gate the error being emitted on whether the `PlaceBase` is a `Local` or not.

It's not possible to copy out of a `static` that contains an `!Freeze` (due to `UnsafeCell` not being `Copy`) or `Drop` type, but if it were, I could demonstrate a bug this allows today.

Reply:
This should get fixed in the old qualification logic as well right? In which case I think it should go in another PR. I'm happy to block this PR until the logic around `static`s is fixed and copy over the changes to this one.

Reply:
It can only be fixed in the old one after we split out the promotability analysis.

Reply:
This is because we promote references to e.g. `static CELL: Cell<i32> = Cell::new(0);` in function bodies?

Reply:
Oops, I meant to post that to the thread about `Rvalue::Ref` not statics.